File: diminishing_circle/diminish.c
#include <stdlib.h>
#include <stdio.h>

struct person {
  struct person* prev;
  struct person* next;
  int index;
};

struct person* setup(int n) {
  struct person* first = malloc(sizeof(struct person));
  first->index = 0;
  struct person* prev = first;
  for(int i = 1; i < n;i++){
    struct person* p = malloc(sizeof(struct person));
    p->prev = prev;
    prev->next = p;
    p->index = i;
    prev = p;
  }
  prev->next = first;
  first->prev = prev;
  return first;
}

struct person* remove_p(struct person* p) {
  p->next->prev = p->prev;
  p->prev->next = p->next;
  struct person* temp = p->prev;
  free(p);
  return temp;
}

struct person* move(int n, struct person* p){
  for(int i = 0; i < n;i++){
    p = p-> next;
  }
  return p;
}

int main(int argc, char* argv[]) {
  int t;
  FILE* file = fopen(argv[1], "r");
  fscanf(file, "%d", &t);
  printf("Number of tests: %d\n", t);
  int n, k;
  for(int i = 0; i < t; i++){
    fscanf(file, "%d %d", &n, &k);
    k++;
    struct person *p = setup(n);
    while(n>1){
      p = move(k%n,p);
      p = remove_p(p);
      n--;
    }
    printf("%d\n", p->index);
  }
}

File: diminishing_circle/diminish.rb
class Person
  attr_accessor :prev, :next, :index

  def initialize index
    @index = index
  end

  def to_s
    return index.to_s
  end

end

#Setup the circle, pass return the first person

def circle_setup n
  first = Person.new 0
  prev = first
  (1..n-1).each do |i|
    p = Person.new(i)
    prev.next = p
    p.prev = prev
    prev = p
  end
  first.prev = prev
  prev.next = first
  return first
end

def remove p
  p.prev.next = p.next
  p.next.prev = p.prev
end

def move n, p
  n.times do
    p = p.next
  end
  return p
end

file = File.open(ARGV[0])

file.readline.to_i.times do
  input = file.readline.split(" ")
  n = input[0].to_i
  k = (input[1].to_i+1)
  p = circle_setup n
  while n > 1
    p = move (k%n), p
    remove p
    n = n-1
  end
  puts p.next
end

